core/src/impl/Kokkos_Utilities.hpp

```diff
@@ -65,6 +65,21 @@ struct identity {
 template <typename T>
 using identity_t = typename identity<T>::type;
 
+struct not_a_type {
+  not_a_type()                  = delete;
+  ~not_a_type()                 = delete;
+  not_a_type(not_a_type const&) = delete;
+  void operator=(not_a_type const&) = delete;
+};
+
+#if defined(__cpp_lib_void_t)
+// since C++17
+using std::void_t;
+#else
+template <class...>
+using void_t = void;
+#endif
```

Comment:
I'm not confident this works on all of the compilers we support (particularly Intel 17)